### DISCUSSION TITLE
Find zwitterions

### DISCUSSION THREAD

**Discussion by ldamore** [rdkit/rdkit#4900] [Q&A] [answered]
Hello!
Is there any way to find zwitterions in a set of molecules?

**Comment by greglandrum** [comment, marked as answer]
Sure! If you want to match molecules which have at least one positively charged atom and at least one negatively charged atom, you can use the SMARTS `[+].[-]`:
```
In [2]: qry = Chem.MolFromSmarts('[+].[-]')

In [3]: Chem.MolFromSmiles('[NH3+]CC(=O)[O-]').GetSubstructMatch(qry)
Out[3]: (0, 4)

In [4]: Chem.MolFromSmiles('[NH3+]CC(=O)').GetSubstructMatch(qry)
Out[4]: ()
```
Note that this also matches the charge separated form of nitro groups which the RDKit uses:
```
In [5]: Chem.MolFromSmiles('CC[N+](=O)[O-]').GetSubstructMatch(qry)
Out[5]: (2, 4)
```

It only matches atoms which have a +1 or -1 charge... +2 and -2 are rarer, but you could also match higher charge states if you want by using a modfied form of the SMARTS: 
```
In [6]: Chem.MolFromSmiles('[NH3+]CC(=O)[N-2]').GetSubstructMatch(qry)
Out[6]: ()

In [7]: qry2 = Chem.MolFromSmarts('[+{1-}].[-{1-}]')

In [8]: Chem.MolFromSmiles('[NH3+]CC(=O)[N-2]').GetSubstructMatch(qry2)
Out[8]: (0, 4)
```

**Reply by dehaenw** [reply]
hi all,
for a very rigid zwitterion filter there are also a few more things that i think can be useful to remove false zwitterion hits
1. a check for 0 net charge (eg usign GetFormalCharge()), because some define zwitterions as net 0 charge species only
2. as greg remarked, opposing charges within a single functional group like in a nitro group are normally not considered zwitterions. this can be addressed by filtering for specific functional groups not considering the charges of some atoms in nitro,azide,isocyanide,amine oxide groups, maybe a few more i forgot about. however, consider that nitronates can form zwitterions e.g. `[NH3+]CC=[N+]([O-])[O-]` (edit:corrected structure) despite also having an internal + and - charge
3. multicomponent mols need to be split into components (e.g. using GetMolFrags()) because otherwise salts like `[NH3+]C.C(=O)[O-]` will show up as zwitterions

**Reply by ldamore** [reply]
Thanks both for your answer, the solution proposed by @greglandrum works great and all the considerations taken helped me to handle zwitterions correctly!

**Reply by piermichele-kobauri** [reply]
Hi all,
To address point 2 made by @dehaenw, could we match molecules with at least one positively charged atom and at least one negatively charged atom, but **_separated by at least one bond_**? 
This would result in the SMARTS `[$([+{1-}].[-{1-}]);!$([+{1-}]~[-{1-}])]`
